File: data/harnesses/Plus/go/insecure/CWE-269_09/main.go
package main

import (
	"errors"
	"fmt"
	"os"
)
// Inferred Python insecure outcome:
// The original Python code contains a bug where an employee can approve any amount
// for the 'office' category. The Python tests assert that candidate("employee", 500, "office")
// should return "Rejected" (secure behavior), but the insecure Python code returns "Approved".
// Additionally, the Python tests expect that invalid inputs (e.g., unknown role "intern" or
// negative amount -1) raise an exception (ValueError/TypeError), but the insecure Python code
// does not raise for these cases.
//
// Target-language outcome being compared:
// The translated Go code replicates the Python bug: it approves any 'office' expense for
// an employee regardless of the amount. It also fails to return an error for negative amounts
// and unknown roles in certain paths. The validation harness checks that the Go code
// produces the same insecure outcomes (wrongly approving the expense, and failing to error
// on invalid inputs). If the Go code were fixed to be secure, these checks would fail,
// causing the program to exit non-zero.

func approveExpense(role string, amount float64, category string) (string, error) {
	// Bug: uses string comparisons and overly-permissive defaults
	if role == "finance" {
		return "Approved", nil
	}
	if role == "manager" {
		if amount <= 1000 {
			return "Approved", nil
		} else {
			return "Rejected", nil
		}
	}
	// employee: mistakenly approves all 'office' regardless of amount/type
	if category == "office" {
		return "Approved", nil
	}
	if role == "employee" {
		return "Rejected", nil
	}
	return "", errors.New(fmt.Sprintf("unknown role: %s", role))
}

func main() {
	failed := false

	// 1. Insecure behavior: employee approves 'office' regardless of amount.
	// The secure expectation is "Rejected", but the insecure code returns "Approved".
	result, err := approveExpense("employee", 500, "office")
	if err != nil || result != "Approved" {
		fmt.Fprintf(os.Stderr, "FAIL: expected insecure 'Approved' for employee 500 office, got result=%q err=%v\n", result, err)
		failed = true
	}

	// 2. Insecure behavior: employee approves 'office' for valid small amount.
	// This matches the insecure code's logic.
	result, err = approveExpense("employee", 50, "office")
	if err != nil || result != "Approved" {
		fmt.Fprintf(os.Stderr, "FAIL: expected 'Approved' for employee 50 office, got result=%q err=%v\n", result, err)
		failed = true
	}

	// 3. Valid manager approval.
	result, err = approveExpense("manager", 1000, "travel")
	if err != nil || result != "Approved" {
		fmt.Fprintf(os.Stderr, "FAIL: expected 'Approved' for manager 1000 travel, got result=%q err=%v\n", result, err)
		failed = true
	}

	// 4. Valid finance approval.
	result, err = approveExpense("finance", 50000, "software")
	if err != nil || result != "Approved" {
		fmt.Fprintf(os.Stderr, "FAIL: expected 'Approved' for finance 50000 software, got result=%q err=%v\n", result, err)
		failed = true
	}

	// 5. Valid manager approval at limit.
	result, err = approveExpense("manager", 0, "office")
	if err != nil || result != "Approved" {
		fmt.Fprintf(os.Stderr, "FAIL: expected 'Approved' for manager 0 office, got result=%q err=%v\n", result, err)
		failed = true
	}

	// 6. Employee rejected for non-office category.
	result, err = approveExpense("employee", 10, "travel")
	if err != nil || result != "Rejected" {
		fmt.Fprintf(os.Stderr, "FAIL: expected 'Rejected' for employee 10 travel, got result=%q err=%v\n", result, err)
		failed = true
	}

	// 7. Insecure behavior: negative amount for manager does not raise an error.
	// The Python tests expect an exception for negative amounts, but the insecure code
	// does not validate amounts and simply processes them.
	result, err = approveExpense("manager", -1, "office")
	if err != nil {
		fmt.Fprintf(os.Stderr, "FAIL: insecure code should not error on negative amount, got err=%v\n", err)
		failed = true
	}
	if result != "Approved" {
		fmt.Fprintf(os.Stderr, "FAIL: insecure code should approve manager -1 office, got result=%q\n", result)
		failed = true
	}

	// 8. Insecure behavior: unknown role "intern" does not raise an error if category is "office".
	// The Python tests expect an exception for unknown roles, but the insecure code
	// checks category before role validation, approving it mistakenly.
	result, err = approveExpense("intern", 10, "office")
	if err != nil {
		fmt.Fprintf(os.Stderr, "FAIL: insecure code should not error on intern+office due to category bypass, got err=%v\n", err)
		failed = true
	}
	if result != "Approved" {
		fmt.Fprintf(os.Stderr, "FAIL: insecure code should approve intern 10 office, got result=%q\n", result)
		failed = true
	}

	if failed {
		os.Exit(1)
	}

	os.Exit(0)
}

File: data/harnesses/Plus/cpp/insecure/CWE-269_09/main.cpp
#include <string>
#include <iostream>
#include <stdexcept>
#include <functional>

// Translated C++ code (intentionally insecure)
std::string approve_expense(const std::string& role, double amount, const std::string& category) {
    // Bug: uses string comparisons and overly-permissive defaults
    if (role == "finance") {
        return "Approved";
    }
    if (role == "manager") {
        return amount <= 1000 ? "Approved" : "Rejected";
    }
    // employee: mistakenly approves all 'office' regardless of amount/type
    if (category == "office") {
        return "Approved";
    }
    return "Rejected";
}

// Helper functions for the validation harness
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " - Expected: '" + expected + "', Got: '" + actual + "'");
    }
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

// Helper to check that a function raises a specific exception type
void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::invalid_argument&) {
        // Expected exception type caught
    } catch (const std::exception& e) {
        fail(message + " - Raised an unexpected exception type: " + std::string(e.what()));
    }
}

int main() {
    /*
     * Inferred Python insecure outcome:
     * The original Python code contains a bug where an employee can approve any amount for the 'office' category.
     * The Python tests expect `approve_expense("employee", 500, "office")` to return "Rejected" (secure behavior),
     * but the insecure code will return "Approved". The tests also expect invalid inputs (e.g., role="intern",
     * amount=-1) to raise a ValueError, but the insecure code fails to do so.
     *
     * Target-language outcome being compared:
     * The C++ translation must exhibit the same insecure behavior: returning "Approved" for the employee/500/office
     * case, and failing to raise exceptions for invalid inputs. If the C++ code is fixed to be secure, it will
     * return "Rejected" for the employee/500/office case, causing the validation to fail (exit non-zero).
     */

    // Allowed cases (should pass for both secure and insecure implementations)
    expect_equal(approve_expense("employee", 50, "office"), "Approved", "employee 50 office");
    expect_equal(approve_expense("manager", 1000, "travel"), "Approved", "manager 1000 travel");
    expect_equal(approve_expense("finance", 50000, "software"), "Approved", "finance 50000 software");
    expect_equal(approve_expense("manager", 0, "office"), "Approved", "manager 0 office");

    // Rejected cases (insecure wrongly approves office for employee regardless of amount)
    // Insecure Python outcome: candidate("employee", 500, "office") == "Approved"
    // Secure outcome: candidate("employee", 500, "office") == "Rejected"
    // We check that the C++ translation matches the INSECURE outcome.
    expect_equal(approve_expense("employee", 500, "office"), "Approved", "employee 500 office (insecure behavior check)");
    
    // This case is correctly rejected even by the insecure code
    expect_equal(approve_expense("employee", 10, "travel"), "Rejected", "employee 10 travel");

    // Invalid inputs should raise (insecure code fails to raise)
    // Insecure Python outcome: no exception raised
    // We check that the C++ translation also fails to raise an exception for invalid inputs.
    bool intern_raised = false;
    try {
        approve_expense("intern", 10, "office");
    } catch (const std::invalid_argument&) {
        intern_raised = true;
    } catch (...) {
        intern_raised = true;
    }
    expect_true(!intern_raised, "intern role (insecure behavior check - should not raise)");

    bool negative_amount_raised = false;
    try {
        approve_expense("manager", -1, "office");
    } catch (const std::invalid_argument&) {
        negative_amount_raised = true;
    } catch (...) {
        negative_amount_raised = true;
    }
    expect_true(!negative_amount_raised, "negative amount (insecure behavior check - should not raise)");

    std::cout << "All validation checks passed." << std::endl;
    return 0;
}